src/Analysis/Elaboration.hs

```diff
@@ -1,87 +1,99 @@
+  pure (maybe (Pure (FreeVariable name)) tvar (envLookup name env) :< Var name)
+elaborate (Fix (Lit b)) = pure (bool :< Lit b)
 elaborate (Fix (If c t e)) = do
-  c' <- elaborate c
+  c' <- check c bool
   t' <- elaborate t
   e' <- elaborate e
   result <- unify (extract t') (extract e')
   pure (result :< If c' t' e')
 elaborate (Fix (Pair fst snd)) = do
   fst' <- elaborate fst
   snd' <- elaborate snd
-  pure (Free (extract fst' :* extract snd') :< Pair fst' snd')
+  pure (extract fst' .* extract snd' :< Pair fst' snd')
 elaborate (Fix (Fst pair)) = do
   t1 <- fresh
   t2 <- fresh
-  pair' <- elaborate pair
-  pairTy <- unify (extract pair') (Free (Free (TVar t1) :* Free (TVar t2)))
-  pure (fromMaybe (Free (TVar t1)) (fstType pairTy) :< Fst pair')
+  pair' <- check pair (tvar t1 .* tvar t2)
+  pure (tvar t1 :< Fst pair')
 elaborate (Fix (Snd pair)) = do
   t1 <- fresh
   t2 <- fresh
-  pair' <- elaborate pair
-  pairTy <- unify (extract pair') (Free (Free (TVar t1) :* Free (TVar t2)))
-  pure (fromMaybe (Free (TVar t2)) (sndType pairTy) :< Snd pair')
+  pair' <- check pair (tvar t1 .* tvar t2)
+  pure (tvar t2 :< Snd pair')
 
+check :: Term -> PartialType Error -> Elab (Cofree Expr (PartialType Error))
+check term ty = do
+  term' <- elaborate term
+  termTy <- unify (extract term') ty
+  pure (termTy :< unwrap term')
 
-unify :: Free Type Error -> Free Type Error -> Elab (Free Type Error)
-unify (Pure err1)   _             = pure (Pure err1)
-unify _             (Pure err2)   = pure (Pure err2)
+
+unify :: PartialType Error -> PartialType Error -> Elab (PartialType Error)
+unify (Pure e1) _         = pure (Pure e1)
+unify _         (Pure e2) = pure (Pure e2)
 unify (Free t1) (Free t2)
-  | t1 == t2  = pure (Free t2)
-  | otherwise = case (t1, t2) of
-    (a1 :-> b1,  a2 :-> b2)  -> (Free .) . (:->) <$> unify a1 a2 <*> unify b1 b2
-    (TVar name1, _)          -> bind name1 (Free t2)
-    (_,          TVar name2) -> bind name2 (Free t1)
-    (t1,         t2)         -> pure (Pure (TypeMismatch t1 t2))
+  | TVar name1 <- t1                   = bind name1 t2
+  |                   TVar name2 <- t2 = bind name2 t1
+  | a1 :-> b1  <- t1, a2 :-> b2  <- t2 = (.->) <$> unify a1 a2 <*> unify b1 b2
+  | a1 :*  b1  <- t1, a2 :*  b2  <- t2 = (.*)  <$> unify a1 a2 <*> unify b1 b2
+  | Bool       <- t1, Bool       <- t2 = pure bool
+  | otherwise = pure (Pure (TypeMismatch t1 t2))
 
-bind :: Name -> Free Type Error -> Elab (Free Type Error)
+bind :: Name -> Type (PartialType Error) -> Elab (PartialType Error)
 bind name ty
+  | TVar name' <- ty, name == name'        = pure (wrap ty)
   | Set.member name (freeTypeVariables ty) = pure (Pure (InfiniteType name ty))
-  | otherwise                              = modify (substExtend name ty) >> pure ty
+  | otherwise                              = do
+    subst <- get
+    maybe (modify (substExtend name (wrap ty)) >> pure (wrap ty)) (unify (wrap ty)) (substLookup name subst)
```

Comment:
Since we aren’t applying the substitution to types as we go, we can revisit the same type variable multiple types. This makes sure we look up and reunify if present.